Add webhook support for Netatmo Cameras

danielperna84 opened the pull request:
## Description:
~This PR updates the netatmo dependency ([pyatmo](https://github.com/jabesq/netatmo-api-python)) to version 1.8. I actually don't know what else has been changed, but I added helper methods to allow support for using webhooks for cameras.~ Update to 1.8 has already been done [here](https://github.com/home-assistant/home-assistant/pull/20938). Webhook support is also what I have added in the component.
In its current form the webhook support only fires events on the bus like such:

Movement:
```
DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event netatmo_webhook_received[L]: home_name=myhome, event_type=movement, webhook_id=mysupersecretwebhookid, camera_id=00:11:22:33:44:55>
```
Person:
```
DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event netatmo_webhook_received[L]: home_name=myhome, message=John seen, event_type=person, id=aaaaaaaa-bbbb-1111-a1a1a1-abcdef123456, webhook_id=mysupersecretwebhookid, name=John, camera_id=00:11:22:33:44:55, is_known=True>
```

I chose not to interact with the sensors this component (optionally) generates because there are no _off_-events for when motion stops or a person hat not been seen for a while. Besides that we also need periodic updates for the required refresh token to be refreshed (pyatmo does that on every update).

I have tested this with a Netatmo Welcome camera, and the existing functionality has not been broken. I also have the weather station with it's various sensors, and they seem to keep working like before as well. For other device types I can't make any statements. If anything breaks, that's due to changes in pyatmo I'm not aware of. But I doubt that there will be any trouble.

I'll take care of the documentation as soon as this PR generally gets accepted.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8463

## Example entry for `configuration.yaml` (if applicable):
```yaml
netatmo:
  api_key: YOUR_CLIENT_ID
  secret_key: YOUR_CLIENT_SECRET
  username: YOUR_USERNAME
  password: YOUR_PASSWORD
  webhook_url: https://hass.example.com/api/webhook/mysupersecretwebhookid
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

danielperna84 committed:
Update Netatmo, add webhook support

danielperna84 committed:
Lint

balloob commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -36,6 +56,7 @@
         vol.Required(CONF_PASSWORD): cv.string,
         vol.Required(CONF_SECRET_KEY): cv.string,
         vol.Required(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_WEBHOOK_URL): cv.string,
```
This is the wrong approach. Netatmo should just generate it's own webhook url and pass it to the integration. The user should not be involved with that at all.

You can generate IDs with:
 -  `hass.components.webhook.async_generate_id`
 -  `hass.components.webhook.async_generate_url`
 -  `hass.components.webhook.async_register`

danielperna84 replied:
But that's now how it works with Netatmo. You have to hard-code an URL on their dev-site. Generating an id/url in HASS would require to update the settings for the "app" every time a new id gets generated. I assume their implementation is targeted at static hosts where they push the data to when the camera generates it. Even though the URL is included when subscribing, it has to be configured in the app as well.

balloob replied:
This would be a lot easier when you use a config entry, that way a webhook ID can be generated and stored and presented to the user when adding the integration.

danielperna84 replied:
Or wait, maybe I'm wrong about this. I'll investigate further.

danielperna84 replied:
You were right. It is possible to use dynamic webhooks. The documentation wasn't clear about this, but after some testing I'm fairly confident dynamic URLs work too.
I did change the config-option to be `webhooks: true`. They should be enabled manually because not everybody exposes hass. And I don't know if subscribing without being able to receive events is a good idea. According to the documentation at least the webhook-function gets banned after 5 times without a 200-response. I don't know how this could impact other parts of the API.

danielperna84 committed:
Switch to autogenerated URL

danielperna84 committed:
Rename event

danielperna84 commented:
I'll probably also change the types of events that are being fired. They should be different depending on what type of event actually happened in my opinion because that also affects what data is available.

danielperna84 committed:
Rename events once more

andersonshatch commented:
This will only support the Welcome camera event types, Presence will send event types of `movement`, `human`, `vehicle`, and `animal`.
Let me know if it will be helpful to provide the body to any of those, I can certainly do that.

danielperna84 commented:
Sure, that would be great.

andersonshatch commented:
Here you go :) https://gist.github.com/andersonshatch/dec84cb83608002e69bfc2759e019c96

danielperna84 committed:
Added Presence events, cleanup

danielperna84 commented:
@andersonshatch 
I have added the events. The possible events on the bus are now:

- netatmo_person
- netatmo_movement
- netatmo_human
- netatmo_animal
- netatmo_vehicle
- netatmo_other

Except from the `netatmo_person` event all events contain the following data:

- event_type
- home_name
- camera_id
- message

For `netatmo_person` there will be an event for each person with the additional data:

- id
- name
- is_known

The `netatmo_other` event will be used for all other possible events we might not know of yet or will be added without our knowledge. I have seen `topology_changed` in the past, which happened when I took it off the power and plugged it back in. There may be other interesting events, and with this solution we are able to monitor them, and eventually add the important ones here as dedicated events.

danielperna84 committed:
Added services to add and drop webhooks, fixed event_other

danielperna84 commented:
I have now also added services to add and drop the webhooks. Adding webhooks is particulary useful because they can get banned. Being able to (re)add the webhook via service removes the need to restart Home Assistant in case the webhook has been banned.
It's also possible to define an URL with the `addwebhook` service. The catch is, that the `handle_webhook` callback won't be subscribed during setup. So people who manually define a webhook as a workaround have to manually work with the received data.

danielperna84 commented:
@andersonshatch 
Could you please try out my latest changes? The automation code I have posted above (now removed) did not work for some reason. I have to admit that I have copied an automation that did something similar and just adjusted it without properly testing it.

What I thought would work:
```yaml
- alias: "Daniel Welcome"
  trigger:
    platform: event
    event_type: netatmo_person
    event_data:
      name: Daniel
```
I do get the matching event. But for some reason the automation won't fire. 😕

danielperna84 committed:
Added face_url for person detection

danielperna84 committed:
Added URLs for Presence camera

andersonshatch commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
+    elif data.get(ATTR_EVENT_TYPE) == EVENT_VEHICLE:
+        hass.bus.async_fire(EVENT_BUS_VEHICLE, published_data)
+        published_data[ATTR_VIGNETTE_URL] = person.get(ATTR_VIGNETTE_URL)
+        published_data[ATTR_SNAPSHOT_URL] = person.get(ATTR_SNAPSHOT_URL)
```
`person.get` should be `data.get` in all `elif` sections:

```Traceback (most recent call last):
  File "/Users/josh/.pyenv/versions/3.6.2/lib/python3.6/site-packages/homeassistant/components/webhook.py", line 78, in async_handle_webhook
    response = await webhook['handler'](hass, webhook_id, request)
  File "/Users/josh/.homeassistant/custom_components/netatmo/__init__.py", line 176, in handle_webhook
    published_data[ATTR_VIGNETTE_URL] = data.get(ATTR_VIGNETTE_URL)
UnboundLocalError: local variable 'person' referenced before assignment```

danielperna84 replied:
Whoops. Copy & paster. 😆

andersonshatch commented:
Other than the one set of changes I annotated, it seems to work okay, though I only tested so far with a dummy event of one type.
I'm confused as to whether I would need to call `netatmo.add_webhook` when I have:
```
netatmo:
  ...
  webhooks: true
```
Is it expected that the webhook is not registered until the service is called? So I would need to create an automation to call this service after hass startup?

danielperna84 committed:
Fix callback

danielperna84 commented:
No, you don't need to call the service. I first added it to be able add and drop during runtime while testing this. In practice it will be irrelevant for me, except if the webhook got banned. In that case you have to go to the dev-panel from Netatmo and unban the app. Afterwards you would need to restart Home Assistant to subscribe again. Using the service lets you do that without restarting Home Assistant.
On top of that the `webhooks: true` only really works with a generic exposed setup. It probably won't work for people with special reverse proxy setups where the auto-generates URL is incorrect. These users then could manually create a working webhook they can work with and add that one by using the service. So essentially it can be ignored by the mayjority of users, but a few wouldn't be excluded just because they expose HASS as usual.

I'll take care of the documentation now.

danielperna84 commented:
@andersonshatch 

Could you please give me a description of what `snapshot_url` and `vignette_url` are for the documentation?

andersonshatch commented:
Sure!
Snapshot_url points to the full frame image captured at the time of that event, whereas vignette_url points to a cropped+zoomed image that focuses on the thing in that event, so if it was an animal event for example, the vignette image will show you just the part the camera believed was an animal.

danielperna84 commented:
@balloob 
To me this seems fine now. Do you agree? Documentation has been taken care of as well.

danielperna84 committed:
Merge branch 'dev' into netatmo

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
             discovery.load_platform(hass, component, DOMAIN, {}, config)
 
+    if config[DOMAIN][CONF_WEBHOOKS]:
+        webhook_id = hass.components.webhook.async_generate_id()
```
We shouldn't call an async function from a sync context. Probably use `hass.add_job` to schedule the async function.

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
 
+    if config[DOMAIN][CONF_WEBHOOKS]:
+        webhook_id = hass.components.webhook.async_generate_id()
+        NETATMO_WEBHOOK_URL = hass.components.webhook.async_generate_url(
```
See above.

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
+        webhook_id = hass.components.webhook.async_generate_id()
+        NETATMO_WEBHOOK_URL = hass.components.webhook.async_generate_url(
+            webhook_id)
+        hass.components.webhook.async_register(
```
See above.

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -61,9 +106,88 @@ def setup(hass, config):
+    """Handle webhook callback."""
+    body = await request.text()
+    try:
+        data = json.loads(body) if body else {}
```
Why not use:
```py
data = await request.json()
```

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -36,16 +74,23 @@
     import pyatmo
 
-    global NETATMO_AUTH
+    global NETATMO_AUTH, NETATMO_WEBHOOK_URL
```
We should not use global. If this needs to be accessed from another module, use hass.data to store it.

MartinHjelmare commented on homeassistant/components/netatmo/__init__.py:
```diff
@@ -106,6 +230,12 @@ def get_camera_type(self, camera=None, home=None, cid=None):
 
+    def get_persons(self):
+        """Gather person data for webhooks."""
+        global NETATMO_PERSONS
```
Don't use global.

MartinHjelmare reviewed (commented):
Please open a new PR where we can address the comments.
